jekyll/_posts/2020-06-04-ray-tracing-in-nim.md

```diff
@@ -0,0 +1,316 @@
+# and auto-convert to float
+converter toF64*(rad: Radians): float64 {.inline.} =
+  float64(rad)
+```
+
+
+### Tracking mutability
+
+When using value types (stack objects, seq and strings), Nim requires an
+explicit `var` when passing parameters or assigning to a variable for
+a parameter to be mutable.
+
+This makes it clear what can or can't move under your feet.
+
+Furthermore, C and C++ developers might be interested to know that
+when passing a large object (over 12 bytes on 32-bit and 24 bytes on 64-bit platforms),
+Nim does the Right Thing™ and will pass by reference by default
+(you can override that on a per-type basis to always pass-by-copy or always by reference).
+This means that function signatures are motivated by mutability concerns and
+uncluttered from performance (and performance is great, as proven by the benchmarks).
+
+
+### Tracking side-effects
+
+Nim functions can be declared `proc` or `func`.
+
+A `func` enforces the absence of side-effect or the code will not compile.
+Some examples of side-effects:
+- (non-deterministic) random functions
+- accessing a global variable
+- multithreading
+- IO
+
+If you have a Heisenbug, it's likely not in a side-effect free function
+(unless you corrupt memory).
+
+This is particularly suited to physics and raytracing computational kernels
+as physics equations are side-effect free.
+
+
+## Compilation times & Meta-programming
+
+Despite all its features, Nim compiles extremely fast with both C and C++ targets.
+In particular template metaprogramming has a very low compilation-time cost.
+
+Nim is the language with the widest support for compile-time metaprogramming,
+including dependently-typed languages.
+
+This is supported by a fast VM that gives Nim the speed of Python at
+compile-time and is used for:
+- writing domain specific languages
+  - state machine generators
+  - seamless multidimensional arrays uses, including Einstein summation
+  - shader generators
+  - SIMD kernel generators
+  - ...
+- compile-time precomputations
+- buffers, vectors, matrices parametrized by integer generics
+
+
+## C and C++ interop
+
+Nim can compile to both C and C++ and seamlessly call libraries written in either languages.
+It can even use CUDA with a bit of configuration to call `nvcc`.
+
+For example, bindings to SFML, from the Nim website features:
+
+<p style="text-align: center;">
+  <img width="715" height="721" src="{{ site.baseurl }}/assets/news/images/mratsim-raytracing/sfml.png">
+</p>
+
+
+## Approachability, productivity & development agility
+
+While Nim offers many advanced features, it does not force you to swim or drown
+on first approach.  
+Actually many have reported that Nim felt like a compiled scripting language.
+
+
+## Caveats
+
+Not all is ponies and rainbows, Nim main issue is that it is a very young language
+with a small ecosystem of supported libraries.
+That said, you can reuse the C and C++ ecosystem (and Javascript as well, since
+Nim can also compile to it).
```

Comment:
```
As an example Trace of Radiance supports video output via MP4 
since v0.2.0 via a [simple 60 lines wrapper](https://github.com/mratsim/trace-of-radiance/blob/v0.2.0/trace_of_radiance/io/mp4.nim#L1-L60)
of an header-only C library.
```